engine/runtime/src/main/java/org/enso/interpreter/runtime/Context.java

```diff
@@ -75,24 +82,40 @@ public Context(Language language, String home, Env environment) {
   /** Perform expensive initialization logic for the context. */
   public void initialize() {
     TruffleFileSystem fs = new TruffleFileSystem();
+    HashMap<String, Package<TruffleFile>> packageMap = new HashMap<>();
     packages = new ArrayList<>();
 
     if (home != null) {
       HomeManager<TruffleFile> homeManager =
           new HomeManager<>(environment.getInternalTruffleFile(home), fs);
-      packages.addAll(homeManager.loadStdLib().collect(Collectors.toList()));
+      packageMap.putAll(
+          homeManager.loadStdLib().collect(Collectors.toMap((Package::name), Function.identity())));
     }
 
     PackageManager<TruffleFile> packageManager = new PackageManager<>(fs);
-
     List<TruffleFile> packagePaths = OptionsHelper.getPackagesPaths(environment);
-    packages.addAll(
-        packagePaths.stream()
-            .map(packageManager::fromDirectory)
-            .map(ScalaConversions::asJava)
-            .filter(Optional::isPresent)
-            .map(Optional::get)
-            .collect(Collectors.toList()));
+
+    // Add user packages one-by-one, shadowing previously added packages. It assumes that the
+    // standard library packages will not clash. In the future, we should be able to disambiguate
+    // packages that clash.
+    for (var packagePath : packagePaths) {
```

Comment:
This should probably warn _in some way_ – doing this silently may result in very surprising behaviors.

Reply:
Oh it does 👌